src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -389,7 +398,8 @@ private void signalIfAllRegisteredInQueue() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size()) {
+        if (registeredElements.size() == liveQueue.size() ||
+                poisonPill != null) {
```

Comment:
If we don't already have a unit test for that, I think having one where we have a `.lock()` block on waiting for the pool to fill back up and have the insertion of poison pill interrupt the wait (presumably throwing an exception) would be a good one to write.